tls/s2n_x509_validator.c

```diff
@@ -368,81 +375,70 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_chain(struct s2n_x509_
         time_t current_time = (time_t)(current_sys_time / 1000000000);
         X509_STORE_CTX_set_time(validator->store_ctx, 0, current_time);
 
-        op_code = X509_verify_cert(validator->store_ctx);
+        int verify_ret = X509_verify_cert(validator->store_ctx);
+        if (verify_ret <= 0) {
+            int ossl_error = X509_STORE_CTX_get_error(validator->store_ctx);
+            switch (ossl_error) {
+                case X509_V_ERR_CERT_HAS_EXPIRED:
+                    RESULT_BAIL(S2N_ERR_CERT_EXPIRED);
+                default:
+                    RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
+            }
+        }
 
-        S2N_ERROR_IF(op_code <= 0, S2N_ERR_CERT_UNTRUSTED);
         validator->state = VALIDATED;
     }
 
     if (conn->actual_protocol_version >= S2N_TLS13) {
-        POSIX_GUARD(s2n_extension_list_process(S2N_EXTENSION_LIST_CERTIFICATE, conn, &first_certificate_extensions));
+        RESULT_GUARD_POSIX(s2n_extension_list_process(S2N_EXTENSION_LIST_CERTIFICATE, conn, &first_certificate_extensions));
     }
 
     *public_key_out = public_key;
 
     /* Reset the old struct, so we don't clean up public_key_out */
     s2n_pkey_zero_init(&public_key);
 
-    return S2N_CERT_OK;
+    return S2N_RESULT_OK;
 }
 
-s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x509_validator *validator,
+S2N_RESULT s2n_x509_validator_validate_cert_stapled_ocsp_response(struct s2n_x509_validator *validator,
         struct s2n_connection *conn, const uint8_t *ocsp_response_raw, uint32_t ocsp_response_length) {
 
     if (validator->skip_cert_validation || !validator->check_stapled_ocsp) {
         validator->state = OCSP_VALIDATED;
-        return S2N_CERT_OK;
+        return S2N_RESULT_OK;
     }
 
-    S2N_ERROR_IF(validator->state != VALIDATED, S2N_ERR_INVALID_CERT_STATE);
+    RESULT_ENSURE(validator->state == VALIDATED, S2N_ERR_INVALID_CERT_STATE);
 
 #if !S2N_OCSP_STAPLING_SUPPORTED
     /* Default to safety */
-    return S2N_CERT_ERR_UNTRUSTED;
+    RESULT_BAIL(S2N_ERR_CERT_UNTRUSTED);
 #else
 
-    OCSP_RESPONSE *ocsp_response = NULL;
-    OCSP_BASICRESP *basic_response = NULL;
-    STACK_OF(X509) *cert_chain = NULL;
-
-    s2n_cert_validation_code ret_val = S2N_CERT_ERR_INVALID;
-
-    if (!ocsp_response_raw) {
-        return ret_val;
-    }
-
-    ocsp_response = d2i_OCSP_RESPONSE(NULL, &ocsp_response_raw, ocsp_response_length);
+    RESULT_GUARD_PTR(ocsp_response_raw);
```

Comment:
Will an error code be set if this is NULL? If not, or we're unsure, it's probably better to:

```suggestion
    RESULT_ENSURE_REF(ocsp_response_raw);
```

Reply:
I fixed both of these in bf827ce6f4c671cf29d94f32af0c52c295060f34.